File: aoc15/day_14/day_14.ml
type reindeer = { speed : int; fly : int; rest : int };;

let parse line =
  let split = String.split_on_char ' ' line in
  let speed = int_of_string (List.nth split 3) in
  let fly = int_of_string (List.nth split 6) in
  let rest = int_of_string (List.nth split 13) in
  { speed; fly; rest }
in
let reindeer = Input.parse_lines parse "input" in

let dst t r =
  let rtt = r.fly + r.rest in
  (t / rtt * r.fly * r.speed) + (min r.fly (t mod rtt) * r.speed)
in

let lead rs t =
  let max_dst = List.fold_left max 0 (List.map (dst t) rs) in
  List.filter (fun r -> dst t r = max_dst) rs
in

let leaders = List.flatten (List.map (lead reindeer) (Utils.int_list 1 2503)) in
let points r = List.length (List.filter (fun _r -> _r = r) leaders) in

let first = List.fold_left max 0 (List.map (dst 2503) reindeer) in
Printf.printf "First: %d\n" first;
let second = List.fold_left max 0 (List.map points reindeer) in
Printf.printf "Second: %d\n" second

File: aoc15/go/day_14/day_14.go
package main

import (
	"bufio"
	"cmp"
	"fmt"
	"os"
	"slices"
	"strconv"
	"strings"
)

type Reindeer struct {
	Speed, Fly, Rest, Dst, Pts int
}

func cmpDst(a, b *Reindeer) int {
	return cmp.Compare(a.Dst, b.Dst)
}

func cmpPts(a, b *Reindeer) int {
	return cmp.Compare(a.Pts, b.Pts)
}

func parse(line string) *Reindeer {
	split := strings.Fields(line)
	speed, _ := strconv.Atoi(split[3])
	fly, _ := strconv.Atoi(split[6])
	rest, _ := strconv.Atoi(split[13])
	return &Reindeer{speed, fly, rest, 0, 0}
}

func main() {
	input, _ := os.Open("input")
	scanner := bufio.NewScanner(input)

	reindeer := []*Reindeer{}
	for scanner.Scan() {
		reindeer = append(reindeer, parse(scanner.Text()))
	}

	for sec := 0; sec < 2503; sec++ {
		for _, r := range reindeer {
			if sec%(r.Fly+r.Rest) < r.Fly {
				r.Dst += r.Speed
			}
		}
		leaderDst := slices.MaxFunc(reindeer, cmpDst).Dst
		for _, r := range reindeer {
			if r.Dst == leaderDst {
				r.Pts++
			}
		}
	}

	fmt.Println("First:", slices.MaxFunc(reindeer, cmpDst).Dst)
	fmt.Println("Second:", slices.MaxFunc(reindeer, cmpPts).Pts)
}
